File: windows/DoNotType.Core.Tests/ToneTests.cs
using DoNotType.Core;
using Xunit;

namespace DoNotType.Core.Tests;

/// <summary>
/// Checks the cues by listening to them rather than by hashing them.
/// </summary>
/// <remarks>
/// The same measurements the Swift suite runs, on this port's own output. That is the point: two
/// suites agreeing on 392 -> 523 and 392 -> 294 is what stops the desktops drifting apart, and a
/// stored copy of the bytes would not do it -- doubles rounded to 16 bits can differ in the last
/// place between platforms without anything being audibly wrong, while a change that made the two
/// cues indistinguishable to a listener could leave the hash intact.
/// </remarks>
public class ToneTests
{
    private const double SampleRate = 48_000.0;

    // The WAV, read back

    private static double[] Samples(byte[] wav)
    {
        var body = DataChunk(wav);
        var samples = new double[body.Length / 2];
        for (var index = 0; index < samples.Length; index++)
        {
            samples[index] = BitConverter.ToInt16(body, index * 2) / 32768.0;
        }
        return samples;
    }

    private static byte[] DataChunk(byte[] wav)
    {
        var cursor = 12;
        while (cursor + 8 <= wav.Length)
        {
            var size = BitConverter.ToInt32(wav, cursor + 4);
            if (wav[cursor] == 'd' && wav[cursor + 1] == 'a'
                && wav[cursor + 2] == 't' && wav[cursor + 3] == 'a')
            {
                var start = cursor + 8;
                return wav[start..Math.Min(start + size, wav.Length)];
            }
            cursor += 8 + size + (size % 2); // chunks are word-aligned
        }
        return [];
    }

    // Measuring a note

    /// <summary>The strongest frequency in a window, to the nearest hertz.</summary>
    /// <remarks>
    /// A Goertzel filter costs one pass per candidate and needs no FFT, which keeps this readable
    /// in all the languages that have to run it. Sweeping whole hertz across the range the cues
    /// occupy is precise enough to tell a fourth from a fifth several times over.
    /// </remarks>
    private static int DominantFrequency(double[] window, int from = 200, int to = 700)
    {
        var best = (Frequency: 0, Power: -1.0);
        for (var candidate = from; candidate <= to; candidate++)
        {
            var power = Goertzel(window, candidate);
            if (power > best.Power)
            {
                best = (candidate, power);
            }
        }
        return best.Frequency;
    }

    private static double Goertzel(double[] window, double frequency)
    {
        var coefficient = 2 * Math.Cos(2 * Math.PI * frequency / SampleRate);
        var previous = 0.0;
        var beforeThat = 0.0;
        foreach (var sample in window)
        {
            var current = sample + (coefficient * previous) - beforeThat;
            beforeThat = previous;
            previous = current;
        }
        return (previous * previous) + (beforeThat * beforeThat)
            - (coefficient * previous * beforeThat);
    }

    /// <summary>The first note alone, then the second once the first has largely decayed.</summary>
    private static (int First, int Second) Notes(byte[] wav)
    {
        var all = Samples(wav);
        var first = all[..(int)(0.12 * SampleRate)];
        var second = all[(int)(0.16 * SampleRate)..(int)(0.30 * SampleRate)];
        return (DominantFrequency(first), DominantFrequency(second));
    }

    // The contract

    /// <summary>G4 up a fourth to C5.</summary>
    [Fact]
    public void StartRisesAFourth()
    {
        var (first, second) = Notes(Tone.Start());
        Assert.InRange(first, 391, 393);
        Assert.InRange(second, 522, 524);
        Assert.True(second > first, "starting rises");
    }

    /// <summary>The same G4 down a fourth to D4.</summary>
    [Fact]
    public void StopFallsAFourth()
    {
        var (first, second) = Notes(Tone.Stop());
        Assert.InRange(first, 391, 393);
        Assert.InRange(second, 293, 295);
        Assert.True(second < first, "stopping falls");
    }

    /// <summary>Both cues open on the same note, which is what makes them a pair.</summary>
    [Fact]
    public void BothCuesShareAnAnchor()
    {
        Assert.Equal(Notes(Tone.Start()).First, Notes(Tone.Stop()).First);
    }

    /// <summary>A cue that outlasts what it reports is still sounding while you talk.</summary>
    [Fact]
    public void EachCueIsUnderHalfASecond()
    {
        foreach (var wav in new[] { Tone.Start(), Tone.Stop() })
        {
            Assert.Equal(0.44, Samples(wav).Length / SampleRate, 2);
        }
    }

    /// <summary>
    /// Quiet on purpose. The floor matters as much as the ceiling: a cue nobody can hear over a
    /// keyboard has failed in the direction that looks like nothing being wrong.
    /// </summary>
    [Fact]
    public void LevelSitsWellBelowFullScale()
    {
        foreach (var wav in new[] { Tone.Start(), Tone.Stop() })
        {
            var loudest = Samples(wav).Max(Math.Abs);
            Assert.InRange(loudest, 0.115, 0.125);
        }
    }

    /// <summary>
    /// <c>SoundPlayer</c> rejects anything it cannot parse, and does it at the point of playback
    /// rather than where the bytes were made.
    /// </summary>
    [Fact]
    public void HeaderIsMono16BitPcmAt48kHz()
    {
        foreach (var wav in new[] { Tone.Start(), Tone.Stop() })
        {
            Assert.Equal("RIFF", System.Text.Encoding.ASCII.GetString(wav, 0, 4));
            Assert.Equal("WAVE", System.Text.Encoding.ASCII.GetString(wav, 8, 4));
            Assert.Equal(1, BitConverter.ToInt16(wav, 20));         // PCM
            Assert.Equal(1, BitConverter.ToInt16(wav, 22));         // mono
            Assert.Equal(48_000, BitConverter.ToInt32(wav, 24));
            Assert.Equal(16, BitConverter.ToInt16(wav, 34));        // bits per sample
            Assert.Equal(wav.Length - 8, BitConverter.ToInt32(wav, 4));
        }
    }

    /// <summary>
    /// The onset is ramped for four milliseconds because a sine wave that starts at full amplitude
    /// is a step change in pressure, and a step is a click.
    /// </summary>
    [Fact]
    public void NoticeableOnsetClickIsRampedAway()
    {
        Assert.InRange(Math.Abs(Samples(Tone.Start())[0]), 0, 0.01);
    }

    /// <summary>
    /// The two desktops play the same cue, so the two ports have to agree sample for sample -- not
    /// merely land on the same notes. Rounding to 16 bits absorbs the last-place differences that
    /// <c>sin</c> and <c>exp</c> are allowed to have between platforms; anything larger is a port
    /// that has drifted. The reference is written by the Swift suite, which owns the original.
    /// </summary>
    [Fact]
    public void MatchesTheSwiftPortSampleForSample()
    {
        var directory = ReferenceDirectory();
        if (directory is null)
        {
            return; // eval/ not reachable from this working directory
        }

        foreach (var (name, wav) in new[] { ("start", Tone.Start()), ("stop", Tone.Stop()) })
        {
            var path = Path.Combine(directory!, $"tone-{name}.wav");
            Assert.True(File.Exists(path), $"missing reference {path}; regenerate with `swift run dnt-eval conformance --write`");

            var reference = Samples(File.ReadAllBytes(path));
            var ours = Samples(wav);
            Assert.Equal(reference.Length, ours.Length);
            for (var index = 0; index < reference.Length; index++)
            {
                // One part in 32768 -- a single step of the 16-bit grid.
                Assert.InRange(Math.Abs(ours[index] - reference[index]), 0, 1.0 / 32768);
            }
        }
    }

    private static string? ReferenceDirectory()
    {
        var directory = new DirectoryInfo(Directory.GetCurrentDirectory());
        for (var depth = 0; depth < 10 && directory is not null; depth++)
        {
            var candidate = Path.Combine(directory.FullName, "eval", "conformance", "tone-start.wav");
            if (File.Exists(candidate))
            {
                return Path.GetDirectoryName(candidate);
            }
            directory = directory.Parent;
        }
        return null;
    }
}

File: Tests/DoNotTypeCoreTests/ToneTests.swift
import Foundation
import XCTest

@testable import DoNotTypeCore

/// Checks the cues by listening to them rather than by hashing them.
///
/// What matters about these two sounds is the pitch of each note and the direction between them —
/// that is the whole reason there are two — so the assertions measure exactly that, with a Goertzel
/// filter over each note in turn. Comparing bytes against a stored copy would pass just as happily
/// on a pair of sounds that had become indistinguishable to a listener, and would fail on a change
/// to the envelope that nobody could hear.
///
/// The Windows port runs the same measurements against its own output. That is the point: the two
/// suites agreeing on 392 → 523 and 392 → 294 is what stops the desktops drifting apart, and CI is
/// the only machine that ever runs both.
final class ToneTests: XCTestCase {
    private let sampleRate = 48_000.0

    // MARK: - Reading a WAV back

    private func samples(_ wav: Data) throws -> [Double] {
        let body = try XCTUnwrap(dataChunk(wav), "no data chunk")
        return stride(from: 0, to: body.count - 1, by: 2).map { index in
            let low = UInt16(body[body.startIndex + index])
            let high = UInt16(body[body.startIndex + index + 1])
            return Double(Int16(bitPattern: low | (high << 8))) / 32768
        }
    }

    private func dataChunk(_ wav: Data) -> Data? {
        var cursor = 12
        while cursor + 8 <= wav.count {
            let id = wav.subdata(in: cursor..<(cursor + 4))
            let size = Int(read32(wav, at: cursor + 4))
            if id == Data("data".utf8) {
                let start = cursor + 8
                return wav.subdata(in: start..<min(start + size, wav.count))
            }
            cursor += 8 + size + (size % 2)
        }
        return nil
    }

    private func read32(_ data: Data, at offset: Int) -> UInt32 {
        (0..<4).reduce(UInt32(0)) { value, byte in
            value | UInt32(data[data.startIndex + offset + byte]) << (8 * UInt32(byte))
        }
    }

    private func read16(_ data: Data, at offset: Int) -> UInt16 {
        UInt16(data[data.startIndex + offset]) | UInt16(data[data.startIndex + offset + 1]) << 8
    }

    // MARK: - Measuring a note

    /// The strongest frequency in a window, to the nearest hertz.
    ///
    /// A Goertzel filter costs one pass per candidate and needs no FFT, which keeps this readable
    /// in all the languages that have to run it. Sweeping whole hertz across the range the cues
    /// occupy is precise enough to tell a fourth from a fifth several times over.
    private func dominantFrequency(_ window: ArraySlice<Double>, from: Int = 200, to: Int = 700)
        -> Int
    {
        var best = (frequency: 0, power: -1.0)
        for candidate in from...to {
            let power = goertzel(window, frequency: Double(candidate))
            if power > best.power { best = (candidate, power) }
        }
        return best.frequency
    }

    private func goertzel(_ window: ArraySlice<Double>, frequency: Double) -> Double {
        let coefficient = 2 * cos(2 * .pi * frequency / sampleRate)
        var previous = 0.0
        var beforeThat = 0.0
        for sample in window {
            let current = sample + coefficient * previous - beforeThat
            beforeThat = previous
            previous = current
        }
        return previous * previous + beforeThat * beforeThat - coefficient * previous * beforeThat
    }

    /// The first note alone, then the second while the first has largely decayed away.
    private func notes(_ wav: Data) throws -> (first: Int, second: Int) {
        let all = try samples(wav)
        let firstWindow = all[0..<Int(0.12 * sampleRate)]
        let secondWindow = all[Int(0.16 * sampleRate)..<Int(0.30 * sampleRate)]
        return (dominantFrequency(firstWindow), dominantFrequency(secondWindow))
    }

    // MARK: - The contract

    /// G4 up a fourth to C5. Tolerance is a hertz either way for the analysis, not the synthesis.
    func testStartRisesAFourth() throws {
        let (first, second) = try notes(Tone.start())
        XCTAssertEqual(first, 392, accuracy: 1, "first note should be G4")
        XCTAssertEqual(second, 523, accuracy: 1, "second note should be C5")
        XCTAssertGreaterThan(second, first, "starting rises")
    }

    /// The same G4 down a fourth to D4.
    func testStopFallsAFourth() throws {
        let (first, second) = try notes(Tone.stop())
        XCTAssertEqual(first, 392, accuracy: 1, "first note should be G4")
        XCTAssertEqual(second, 294, accuracy: 1, "second note should be D4")
        XCTAssertLessThan(second, first, "stopping falls")
    }

    /// Both cues open on the same note, which is what makes them a pair rather than two sounds.
    func testBothCuesShareAnAnchor() throws {
        XCTAssertEqual(try notes(Tone.start()).first, try notes(Tone.stop()).first)
    }

    /// A cue that outlasts what it reports is worse than none: it is still sounding while you talk.
    func testEachCueIsUnderHalfASecond() throws {
        for wav in [Tone.start(), Tone.stop()] {
            let duration = Double(try samples(wav).count) / sampleRate
            XCTAssertEqual(duration, 0.44, accuracy: 0.01)
        }
    }

    /// Quiet on purpose — see `Tone.peak`. The floor matters as much as the ceiling: a cue nobody
    /// can hear over a keyboard has failed in the direction that looks like nothing being wrong.
    func testLevelSitsWellBelowFullScale() throws {
        for wav in [Tone.start(), Tone.stop()] {
            let loudest = try samples(wav).map(abs).max() ?? 0
            XCTAssertEqual(loudest, 0.12, accuracy: 0.005)
        }
    }

    /// `NSSound` and `SoundPlayer` both reject anything they cannot parse, and both do it silently
    /// at the point of playback rather than where the bytes were made.
    func testHeaderIsMono16BitPcmAt48kHz() throws {
        for wav in [Tone.start(), Tone.stop()] {
            XCTAssertEqual(wav.subdata(in: 0..<4), Data("RIFF".utf8))
            XCTAssertEqual(wav.subdata(in: 8..<12), Data("WAVE".utf8))
            XCTAssertEqual(read16(wav, at: 20), 1, "PCM")
            XCTAssertEqual(read16(wav, at: 22), 1, "mono")
            XCTAssertEqual(read32(wav, at: 24), 48_000)
            XCTAssertEqual(read16(wav, at: 34), 16, "bits per sample")
            XCTAssertEqual(Int(read32(wav, at: 4)), wav.count - 8, "RIFF size counts what follows")
        }
    }

    /// The onset is ramped for four milliseconds because a sine wave that starts at full amplitude
    /// is a step change in pressure, and a step is a click. This is that ramp.
    func testNoticeableOnsetClickIsRampedAway() throws {
        let opening = try samples(Tone.start())[0]
        XCTAssertLessThan(abs(opening), 0.01)
    }
}
